packages/network-controller/tests/NetworkController.provider.test.ts

```diff
@@ -0,0 +1,894 @@
+      .reply(200, {
+        id: 1,
+        jsonrpc: '2.0',
+        result: '0x1',
+      })
+      .post('/', {
+        id: 1,
+        jsonrpc: '2.0',
+        method: rpcMethod,
+        params: [],
+      })
+      .reply(200, {
+        id: 1,
+        jsonrpc: '2.0',
+        result: 'ok',
+      });
+
+    await withController(
+      {
+        isRpcFailoverEnabled: true,
+        state: {
+          networkConfigurationsByChainId: {
+            '0x1337': buildCustomNetworkConfiguration({
+              chainId: '0x1337',
+              name: 'Test Network',
+              rpcEndpoints: [
+                buildCustomRpcEndpoint({
+                  networkClientId,
+                  url: primaryEndpointUrl,
+                  failoverUrls: [secondaryEndpointUrl],
+                }),
+              ],
+            }),
+          },
+          networksMetadata: {
+            [networkClientId]: {
+              EIPS: {},
+              status: NetworkStatus.Unknown,
+            },
+          },
+          selectedNetworkClientId: networkClientId,
+        },
+      },
+      async ({ controller, messenger }) => {
+        messenger.subscribe('NetworkController:rpcEndpointRetried', () => {
+          clock.next();
+        });
+        const stateChangeListener = jest.fn();
+        messenger.subscribe(
+          'NetworkController:stateChange',
+          stateChangeListener,
+        );
+        const { provider } = controller.getNetworkClientById(networkClientId);
+        const request = {
+          id: 1,
+          jsonrpc: '2.0' as const,
+          method: rpcMethod,
+          params: [],
+        };
+        const expectedError = 'RPC endpoint not found or unavailable';
+
+        // Hit the primary, run out of retries.
+        await expect(provider.request(request)).rejects.toThrow(expectedError);
+        // Hit the primary, run out of retries.
+        await expect(provider.request(request)).rejects.toThrow(expectedError);
+        // Hit the primary, break the circuit, fail over to the secondary.
+        await provider.request(request);
+
+        expect(stateChangeListener).toHaveBeenCalledTimes(2);
+        expect(stateChangeListener).toHaveBeenNthCalledWith(
+          1,
+          expect.any(Object),
+          [
+            {
+              op: 'replace',
+              path: ['networksMetadata', 'AAAA-AAAA-AAAA-AAAA', 'status'],
+              value: 'degraded',
+            },
+          ],
+        );
+        expect(stateChangeListener).toHaveBeenNthCalledWith(
+          2,
+          expect.any(Object),
+          [
+            {
+              op: 'replace',
+              path: ['networksMetadata', 'AAAA-AAAA-AAAA-AAAA', 'status'],
+              value: 'available',
+            },
+          ],
+        );
+      },
+    );
+  });
+
+  it('does not transition the status of a network client from "degraded" the first time a failover is activated but it does not return a 2xx response', async () => {
```

Comment:
Nit: I found this difficult to understand because of the awkward "but" and the double-negative. Here's an attempt at simplifying:

```suggestion
  it('does not transition the status of a network client from "degraded" the first time a failover is activated if it returns a non-2xx response', async () => {
```